### PR TITLE
Ensure that "require(i < 2**255);" parses correctly

### PR DESCRIPTION
**Pull request by seanyoung**
Found with uniswap v3. This change also includes a little refactoring, and now Expression::NotEqual is generated by sema.

Before this change, 

```solidity
require(i < 2**255);
```
Gave the error: value 53919893334301279589334030174039261347274288845081144962207220498432 does not fit into type uint8.

This is because with `ResolveTo::Unknown` the integers default to the smallest size possible.

### CODE REVIEW

File: src/sema/expression/arithmetic.rs [hunk 1]
```diff
@@ -539,11 +556,130 @@ pub(super) fn equal(
...
+
+    // Comparing stringliteral against stringliteral
+    if let (Expression::BytesLiteral { value: l, .. }, Expression::BytesLiteral { value: r, .. }) =
+        (&left, &right)
+    {
+        return Ok(Expression::BoolLiteral {
+            loc: *loc,
+            value: l != r,
+        });
+    }
```
[the remark this answers is not in the document]
Comment (seanyoung) [reply src/sema/expression/arithmetic.rs:626]: Good point

File: src/emit/expression.rs [hunk 2]
```diff
@@ -785,12 +785,41 @@ pub(super) fn expression<'a, T: TargetRuntime<'a> + ?Sized>(
...
+            if left.ty().is_address() {
+                let mut res = bin.context.bool_type().const_int(0, false);
+                let left = expression(target, bin, left, vartab, function, ns).into_array_value();
+                let right = expression(target, bin, right, vartab, function, ns).into_array_value();
 
-            bin.builder
-                .build_int_compare(IntPredicate::NE, left, right, "")
-                .into()
+                // TODO: Address should be passed around as pointer. Once this is done, we can replace
+                // this with a call to address_equal()
```
[the remark this answers is not in the document]
Comment (seanyoung) [reply src/emit/expression.rs:794]: comparing two `uint256` looks nice in llvm but not in BPF.
 - Does not work on address of non-32-bytes
 - Better to compare 64 bits at a time

[the remark this answers is not in the document]
Comment (seanyoung) [reply src/emit/expression.rs:794]: address is an array, you can't just type pun an array into an i256 in llvm

Having said that, the codegen is awful on BPF.

File: src/sema/expression/arithmetic.rs [hunk 3]
```diff
@@ -539,11 +556,130 @@ pub(super) fn equal(
...
+
+fn is_string_equal(
+    loc: &pt::Loc,
+    left: &Expression,
+    left_type: &Type,
+    right: &Expression,
+    right_type: &Type,
+    ns: &Namespace,
+    diagnostics: &mut Diagnostics,
+) -> Result<Option<Expression>, ()> {
```
[the remark this answers is not in the document]
Comment (seanyoung) [reply src/sema/expression/arithmetic.rs:677]: I've added a comment. I don't think the return type is ugly.

File: src/sema/expression/resolve_expression.rs [hunk 4]
```diff
@@ -722,3 +432,344 @@ pub fn expression(
...
+        pt::UserDefinedOperator::BitwiseNot,
+        diagnostics,
+        ns,
+    ) {
+        return Ok(expr);
+    }
+
+    let expr_ty = expr.ty();
+
+    get_int_length(&expr_ty, loc, true, ns, diagnostics)?;
```
[the remark this answers is not in the document]
Comment (seanyoung) [reply src/sema/expression/resolve_expression.rs:461]: I've added a comment and renamed the function.

File: src/sema/expression/arithmetic.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (seanyoung) [reply src/sema/expression/arithmetic.rs]: Fixed

**Review by xermicus — APPROVED** [review]
LGTM

File: src/codegen/constant_folding.rs
Comment (LucasSte) [inline src/codegen/constant_folding.rs:1]: There are no tests for the changes in this file.

**Pull request merged**